pkg/plugins/runtime/k8s/controllers/cni_taint_controller.go

```diff
@@ -0,0 +1,197 @@
+}
+
+func (r *CniNodeTaintReconciler) updateTaints(ctx context.Context, log logr.Logger, node *kube_core.Node, pods []kube_core.Pod) error {
+	if hasTaint(node) {
+		if hasCniPodRunning(log, pods) {
+			log.Info("has cni pod running and taint")
+			return r.untaintNode(ctx, log, node)
+		} else {
+			log.Info("has no cni pod running and taint")
+			return nil
+		}
+	} else {
+		if hasCniPodRunning(log, pods) {
+			log.Info("has cni pod running and no taint")
+			return nil
+		} else {
+			log.Info("has no cni pod running and no taint")
+			return r.taintNode(ctx, log, node)
+		}
+	}
+}
+
+func (r *CniNodeTaintReconciler) untaintNode(ctx context.Context, log logr.Logger, node *kube_core.Node) error {
+	taintIndex := slices.IndexFunc(node.Spec.Taints, func(taint kube_core.Taint) bool {
+		return taint.Key == nodeReadinessTaintKey && taint.Effect == kube_core.TaintEffectNoSchedule
+	})
+
+	if taintIndex >= 0 {
+		node.Spec.Taints = removeTaint(node.Spec.Taints, taintIndex)
+	}
+
+	err := r.Client.Update(ctx, node)
+	if err == nil {
+		log.Info("removed the taint from node")
+	}
+	return err
+}
+
+func removeTaint(s []kube_core.Taint, index int) []kube_core.Taint {
+	return append(s[:index], s[index+1:]...)
+}
+
+func (r *CniNodeTaintReconciler) taintNode(ctx context.Context, log logr.Logger, node *kube_core.Node) error {
+	node.Spec.Taints = append(node.Spec.Taints, kube_core.Taint{
+		Key:    nodeReadinessTaintKey,
+		Effect: kube_core.TaintEffectNoSchedule,
+	})
+
+	err := r.Client.Update(ctx, node)
+	if err == nil {
+		log.Info("added taint to node")
+	}
+	return err
+}
+
+func hasTaint(node *kube_core.Node) bool {
+	foundTaint := false
+	for _, taint := range node.Spec.Taints {
+		if taint.Key == nodeReadinessTaintKey && taint.Effect == kube_core.TaintEffectNoSchedule {
+			foundTaint = true
+		}
+	}
+	return foundTaint
+}
+
+func hasCniPodRunning(log logr.Logger, pods []kube_core.Pod) bool {
+	for _, pod := range pods {
+		if isCniPod(pod) && pod.Status.Phase == kube_core.PodRunning {
+			for _, condition := range pod.Status.Conditions {
+				if condition.Type == kube_core.PodReady && condition.Status == kube_core.ConditionTrue {
+					log.V(1).Info("pod has kuma-cni-node running and ready", "pod", pod.Name, "status", pod.Status)
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+func isCniPod(pod kube_core.Pod) bool {
+	value, ok := pod.Labels[parentAppLabel]
+	return ok && value == metadata.KumaCNI
+}
+
+func (r *CniNodeTaintReconciler) SetupWithManager(mgr kube_ctrl.Manager) error {
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &kube_core.Pod{}, nodeIndexField, func(obj kube_client.Object) []string {
+		pod := obj.(*kube_core.Pod)
+		return []string{pod.Spec.NodeName}
+	}); err != nil {
+		return err
+	}
+
+	return kube_ctrl.NewControllerManagedBy(mgr).
+		For(&kube_core.Node{}, builder.WithPredicates(nodeEvents)).
+		// check this is necessary
+		Watches(
+			&kube_source.Kind{Type: &kube_core.Pod{}},
```

Comment:
What are the reasons we need to watch on pods? (That's a lot more events that just node :) )

Reply:
If we don't watch pods how would we know that the `kuma-cni` pod is up and running? We could periodically check it but I defaulted to what similar taint controllers do.